[gym][snapshot] fix gym `use_system_scm` option and add `use_system_scm` option to snapshot

rogerluan opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
Resolves #17767

### Description
This PR is a follow up of #17644, that introduced the use_system_scm parameter to `gym` and `scan`.
The issue was that that option was not being forwarded to (or read by, actually) `xcodebuild`'s ``-showBuildSettings` command.

Prior to this PR:

```
$ be fastlane run gym use_system_scm:true 
[✔] 🚀 
[18:13:37]: -----------------
[18:13:37]: --- Step: gym ---
[18:13:37]: -----------------
Select Scheme: 
1. redacted1
2. redacted2
3. redacted3
4. redacted4
5. redacted5
6. redacted6
7. redacted7
8. redacted8
9. redacted9
?  6
[18:13:39]: Resolving Swift Package Manager dependencies...
[18:13:39]: $ xcodebuild -resolvePackageDependencies -workspace ./redacted.xcworkspace -scheme redacted6
```

After this PR:

```
$ be fastlane run gym use_system_scm:true
[✔] 🚀 
[18:14:52]: -----------------
[18:14:52]: --- Step: gym ---
[18:14:52]: -----------------
Select Scheme: 
1. redacted1
2. redacted2
3. redacted3
4. redacted4
5. redacted5
6. redacted6
7. redacted7
8. redacted8
9. redacted9
?  6
[18:14:54]: Resolving Swift Package Manager dependencies...
[18:14:54]: $ xcodebuild -resolvePackageDependencies -workspace ./redacted.xcworkspace -scheme redacted6 -scmProvider system
```

This PR also adds `use_system_scm` option to snapshot, to fix specs.

### Testing Steps

To test this branch, modify your Gemfile as:

```ruby
gem "fastlane", :git => "https://github.com/fastlane/fastlane.git", :branch => "rogerluan-fix-gym-use-system-scm"
```

Then you can test it by running gym/build_app command using the `use_system_scm` option, such as:

`bundle exec fastlane run gym use_system_scm:true`

And verifying that it appends `-scmProvider system` to the end of the xcodebuild command.

rogerluan committed:
Fix gym -showBuildSettings command when providing use_system_scm option.
- Also add use_system_scm option to snapshot to fix specs.

rogerluan committed:
Merge branch 'master' into rogerluan-fix-gym-use-system-scm

ainame reviewed (dismissed):
Looks good to me👍 Nice catch!

felixscheinost commented:
When will this be merged?

plu commented:
Looks great, looking forward to see this merged and released. What's holding this back?

This option solves CI working with private SwiftPM packages referenced via `git@...`

plu commented:
I've just tried the branch, it partially looks good:

```
[15:55:25]: $ xcodebuild -resolvePackageDependencies -workspace eBay\ Kleinanzeigen.xcworkspace -scheme eBay\ Kleinanzeigen -derivedDataPath /Users/jenkins/Library/Caches/ios-client/DerivedData/Test -scmProvider system
```

But then later on it generates a wrong command:

```
[15:55:50]: $ set -o pipefail && env NSUnbufferedIO=YES xcodebuild -workspace eBay\ Kleinanzeigen.xcworkspace -scheme eBay\ Kleinanzeigen -derivedDataPath /Users/jenkins/Library/Caches/ios-client/DerivedData/Test -scmProvider system -scmProvider system -destination 'platform=iOS Simulator,id=8744F718-E0FB-4530-8BE5-A847CD631BD7' -resultBundlePath '/Users/jenkins/workspace/iOS_main_pipeline_PR-3155/test-reports/unit/eBay Kleinanzeigen.xcresult' build test | tee '/Users/jenkins/workspace/iOS_main_pipeline_PR-3155/test-reports/unit/Kleinanzeigen-eBay Kleinanzeigen.log' | xcpretty  --report html --output '/Users/jenkins/workspace/iOS_main_pipeline_PR-3155/test-reports/unit/report.html' --report junit --output '/Users/jenkins/workspace/iOS_main_pipeline_PR-3155/test-reports/unit/report.junit' --report junit --output '/var/folders/py/hqhn8n_n0bn3v7jmr9j4631m0000gl/T/junit_report20210112-71370-z590it' 
[15:55:50]: ▸ Loading...
[15:55:51]: ▸ xcodebuild: error: option '-scmProvider' may only be provided once
```

Adding `-scmProvider system` twice.

ainame commented:
@rogerluan Any idea on that^?

rogerluan commented:
@plu thanks for the feedback! Could you provide the command you used and the full output of the command (running `--verbose`)? 🙏

plu commented:
Please check out #17957, it target the branch of this PR. I've added some failing tests there and a fix for it.

plu committed:
Fix duplicate use_system_scm option in #17832 (#17957)

* Bump rspec version

This will allow usage of `include("foo").once`

* Add failing test

      1) Scan Scan::TestCommandGenerator uses system scm options exactly once
         Failure/Error: expect(result).to include("-scmProvider system").once

           expected ["set -o pipefail &&", "env NSUnbufferedIO=YES xcodebuild", "-scheme app", "-project ./scan/examples/...it --output '/var/folders/qz/91qlx7gn30l6rfcwjfzsv_ph0000gn/T/junit_report20210113-97481-17c5v9n' "] to include "-scmProvider system" once but it is included twice

* Deduplicate scmProvider option

* Fix build command generator

* Remove extra blank line

plu commented:
@ainame could you have another look, please? The changes/fixes from the other PR have landed here now. I have integrated the changes on our CI, and I can confirm that "real world" usage of `gym` and `scan` are working there fine now, using `use_system_scm: true`.

ainame commented on fastlane.gemspec:
```diff
@@ -102,7 +102,7 @@ Gem::Specification.new do |spec|
   # Development only
   spec.add_development_dependency('rake')
-  spec.add_development_dependency('rspec', '~> 3.9.0')
+  spec.add_development_dependency('rspec', '~> 3.10.0')
```
Note: This is to use `include(...).once` matcher.

https://github.com/fastlane/fastlane/pull/17957#discussion_r557052834

ainame commented on gym/lib/gym/generators/build_command_generator.rb:
```diff
@@ -39,7 +39,9 @@ def options
         options << "-archivePath #{archive_path.shellescape}" unless config[:skip_archive]
         options << "-resultBundlePath '#{result_bundle_path}'" if config[:result_bundle]
-        options << "-scmProvider system" if config[:use_system_scm]
+        if config[:use_system_scm] && !options.include?("-scmProvider system")
```
I think we should improve the entire flow to compose xcodebuild options to no have duplicated options but this is fine for now; e.g. Clarify what `project_path_array` brings and use Hash to manage options and convert them to real command argument eventually.

Thank you for taking care of this!

ainame reviewed (approved):
LGTM again🚀  @joshdholtz will double-check and merge when he can!

joshdholtz reviewed (approved):
This looks great! Thank you so much for making this fix ❤
